### DISCUSSION TITLE
Vetting an object that can be a subset of another object

### DISCUSSION THREAD

**Discussion by MikkelHJuul** [cue-lang/cue#1163] [General] [answered]
When vetting an object the vet fails validating an object that is a subset of another object in an or-list of possible structures.

if `#A` is a subset of `#B` then 
`obj: #A | #B` fails validating an object describing `#A`
with message: `incomplete value {...#A<exploded - values replaced>} | {...#B<exploded - values replaced>}`

Is this the correct behavior, or a missing feature?

this sprang from the discussion: https://cuelang.slack.com/archives/CLT3ULF6C/p1626443556244700

**Comment by myitcv** [comment, marked as answer]
@MikkelHJuul please can you provide an example of what is failing?

**Reply by MikkelHJuul** [reply]
You can see an example in https://github.com/cue-lang/cue/discussions/1164

I can see/hear from @verdverm(as he also noted in the slack thread), that this is the expected behavior, but as I see it specifically for the case of vet, and e.g. validation of a specification, where the object must comply to `#A` or `#B` (fully). It pushed me to allow any subset of `#B` that contains `#A`, which is quite unsatisfactory.

So this would be a feature request? I can understand the cue eval-behavior, but for `vet` there should maybe be a step more that allows for this validation (if on object `X` comply to `#A`, and `X` only have the elements of `#A` (and no other elements); vet should pass). (is this a performance-consideration?)

**Reply by myitcv** [reply]
> I can understand the cue eval-behavior, but for `vet` there should maybe be a step more that allows for this validation (if on object `X` comply to `#A`, and `X` only have the elements of `#A` (and no other elements); vet should pass). (is this a performance-consideration?)

Thanks. I suspected you were hinting at this sort of mode for `vet`. But I wonder whether actually you are running up against a need for required fields #822. As @verdverm points out, it's generally a bit of a smell that something is not quite right if you have a disjunction where some elements of the disjunction [subsume](https://tip.cuelang.org/docs/references/spec/#values-1) others. 


> and no other elements

This is what closedness is for. Hence I suspect the piece we are missing here, is that in the type `#B` (if we assume that `#B` subsumes `#A`) there are required fields. Is that the case?

**Reply by MikkelHJuul** [reply]
it is specifically that the one object `#A` must have all it's fields set, and the other object `#B` must have all it's fields set, but both settings are valid configurations, nothing in between is valid.

all fields of both objects are required, but `#B` subsumes `#A`

**Reply by MikkelHJuul** [reply]
It is unclear if the proposal would fix this, maybe something like:
```CUE
obj: {
    numexist(=1, a) |  numexist(=2, a, b)
    a: string
    b: string
}
```
which seems a bit like a hack (if it is even valid, I'm also not sure how the bulitin `numexist` work)

**Reply by myitcv** [reply]
> all fields of both objects are required, but `#B` subsumes `#A`

If it is the case that all fields in both objects are required, then `#B` does not subsume `#A` AFAIU. Because for that to be the case, a value `v: #A` would also have to satisfy `#B`. And that cannot be the case under the required fields proposal (unless `#B` and `#A` are identical). Because any required field that `#B` defines beyond `#A` would not be permitted by `#A` because of closedness, and could not be omitted in a check against `#B` because it is required. 

A very basic example of what I understand from your situation would be as follows:

```
#A: {
	name!: string
}

#B: {
	#A
	age!: int
}
```

Both `name` and `age` are marked as required, so a value `{name: "cueckoo"}` is unambiguously of type `#A`, and `{name: "gopher", age: 42}` is unambiguously of type `#B`.

**Reply by MikkelHJuul** [reply]
exactly - now it almost sounds like I am supposed to be able to do this, but I am not. I always get the error:  `incomplete value {...#A<exploded - values replaced>} | {...#B<exploded - values replaced>}` 

fx
```CUE
#A: a: string
#B: a: string
#B: b: string

a: #A|#B
```
```shell
> echo '{"a":"asd"}' >> t.json
> cue vet t.json test.cue -d a
incomplete value {a:"asd"} | {a:"asd",b:string}
```
or 
```CUE
close({
   a: string
}) | close({
   a: string
   b: string
})
```
```shell
> cue vet t.json test.cue
incomplete value {a:"asd"} | {a:"asd",b:string}
```
or am I doing something incorrect?

**Reply by myitcv** [reply]
> or am I doing something incorrect?

You are not doing anything incorrect at all. It's just that you need #822 to be implemented in order to achieve what you want 😄 

Notice in my example above, the use of `!` in the declaration of the two fields. That is the proposed syntax for specifying that a field is required. And it is those required semantics that allow such a disjunction to be resolved unambiguously. Without those semantics, i.e. today, you cannot do it.

**Reply by MikkelHJuul** [reply]
okay, great! thanks

**Comment by verdverm** [comment]
This is expected behavior when the fields of one struct are a strict subset of another struct in a disjunction. There is no way for CUE to decide which is "correct." Each option to a disjunction needs to be uniquely differentiated. When you find that one is a subset of another, there may be a different way to organize things depending am what you are trying to achieve.
